rclpy/rclpy/action_client.py

```diff
@@ -0,0 +1,498 @@
+    ):
+        """
+        Constructor.
+
+        :param node: The ROS node to add the action client to.
+        :param action_type: Type of the action.
+        :param action_name: Name of the action.
+            Used as part of the underlying topic and service names.
+        :param callback_group: Callback group to add the action client to.
+            If None, then the node's default callback group is used.
+        :param goal_service_qos_profile: QoS profile for the goal service.
+        :param result_service_qos_profile: QoS profile for the result service.
+        :param cancel_service_qos_profile: QoS profile for the cancel service.
+        :param feedback_sub_qos_profile: QoS profile for the feedback subscriber.
+        :param status_sub_qos_profile: QoS profile for the status subscriber.
+        """
+        if callback_group is None:
+            callback_group = node.default_callback_group
+
+        super().__init__(callback_group)
+
+        # Import the typesupport for the action module if not already done
+        check_for_type_support(action_type)
+        self._node = node
+        self._action_type = action_type
+        self._client_handle = _rclpy_action.rclpy_action_create_client(
+            node.handle,
+            action_type,
+            action_name,
+            goal_service_qos_profile.get_c_qos_profile(),
+            result_service_qos_profile.get_c_qos_profile(),
+            cancel_service_qos_profile.get_c_qos_profile(),
+            feedback_sub_qos_profile.get_c_qos_profile(),
+            status_sub_qos_profile.get_c_qos_profile()
+        )
+
+        self._is_ready = False
+        self._goal_handles = {}
+        self._pending_goal_requests = {}
+        self._sequence_number_to_goal_id = {}  # goal request sequence number
+        self._pending_cancel_requests = {}
+        self._pending_result_requests = {}
+        self._feedback_callbacks = {}
+
+        callback_group.add_entity(self)
+        self._node.add_waitable(self)
+
+    def _generate_random_uuid(self):
+        return UUID(uuid=list(uuid.uuid4().bytes))
+
+    def _remove_pending_request(self, future, pending_requests):
+        """
+        Remove a future from the list of pending requests.
+
+        This prevents a future from receiving a request and executing its done callbacks.
+        :param future: a future returned from one of :meth:`send_goal_async`,
+            :meth:`cancel_goal_async`, or :meth:`get_result_async`.
+        :type future: rclpy.task.Future
+        :param pending_requests: The list of pending requests.
+        :type pending_requests: dict
+        :return: The sequence number associated with the removed future, or
+            None if the future was not found in the list.
+        """
+        for seq, req_future in pending_requests.items():
+            if future == req_future:
+                try:
+                    del pending_requests[seq]
```

Comment:
No explicit tests, but I tried a test sending multiple goals concurrently and still works. It would raise an exception, but the iteration stops on the first deletion because of the return statement below. Anyways, it's probably  safer to cast the items to a list as suggested in the SO post.

We might consider doing the same for the service client code as well:

https://github.com/ros2/rclpy/blob/0fafaec6b3764654413d0ce37839d7c7178245af/rclpy/rclpy/client.py#L70-L76

Reply:
Cast to list: 04337c9